core/crypto/sm3.c

```diff
@@ -0,0 +1,276 @@
+	ctx->state[3] = 0xDA8A0600;
+	ctx->state[4] = 0xA96F30BC;
+	ctx->state[5] = 0x163138AA;
+	ctx->state[6] = 0xE38DEE4D;
+	ctx->state[7] = 0xB0FB0E4E;
+}
+
+static void sm3_process(struct sm3_context *ctx, const uint8_t data[64])
+{
+	uint32_t SS1, SS2, TT1, TT2, W[68], W1[64];
+	uint32_t A, B, C, D, E, F, G, H;
+	uint32_t T[64];
+	uint32_t Temp1, Temp2, Temp3, Temp4, Temp5;
+	int j;
+
+	for (j = 0; j < 16; j++)
+		T[j] = 0x79CC4519;
+	for (j = 16; j < 64; j++)
+		T[j] = 0x7A879D8A;
+
+	W[0] = get_be32(data);
+	W[1] = get_be32(data + 4);
+	W[2] = get_be32(data + 8);
+	W[3] = get_be32(data + 12);
+	W[4] = get_be32(data + 16);
+	W[5] = get_be32(data + 20);
+	W[6] = get_be32(data + 24);
+	W[7] = get_be32(data + 28);
+	W[8] = get_be32(data + 32);
+	W[9] = get_be32(data + 36);
+	W[10] = get_be32(data + 40);
+	W[11] = get_be32(data + 44);
+	W[12] = get_be32(data + 48);
+	W[13] = get_be32(data + 52);
+	W[14] = get_be32(data + 56);
+	W[15] = get_be32(data + 60);
+
+#define FF0(x, y, z)	((x) ^ (y) ^ (z))
+#define FF1(x, y, z)	(((x) & (y)) | ((x) & (z)) | ((y) & (z)))
+
+#define GG0(x, y, z)	((x) ^ (y) ^ (z))
+#define GG1(x, y, z)	(((x) & (y)) | ((~(x)) & (z)))
+
+#define SHL(x, n)	(((x) & 0xFFFFFFFF) << (n))
+#define ROTL(x, n)	(SHL((x), (n)) | ((x) >> (32 - (n))))
+
+#define P0(x)	((x) ^ ROTL((x), 9) ^ ROTL((x), 17))
+#define P1(x)	((x) ^ ROTL((x), 15) ^ ROTL((x), 23))
+
+	for (j = 16; j < 68; j++) {
+		/*
+		 * W[j] = P1( W[j-16] ^ W[j-9] ^ ROTL(W[j-3],15)) ^
+		 *        ROTL(W[j - 13],7 ) ^ W[j-6];
+		 */
+
+		Temp1 = W[j - 16] ^ W[j - 9];
+		Temp2 = ROTL(W[j - 3], 15);
+		Temp3 = Temp1 ^ Temp2;
+		Temp4 = P1(Temp3);
+		Temp5 =  ROTL(W[j - 13], 7) ^ W[j - 6];
+		W[j] = Temp4 ^ Temp5;
+	}
+
+	for (j =  0; j < 64; j++)
+		W1[j] = W[j] ^ W[j + 4];
+
+	A = ctx->state[0];
+	B = ctx->state[1];
+	C = ctx->state[2];
+	D = ctx->state[3];
+	E = ctx->state[4];
+	F = ctx->state[5];
+	G = ctx->state[6];
+	H = ctx->state[7];
+
+	for (j = 0; j < 16; j++) {
+		SS1 = ROTL((ROTL(A, 12) + E + ROTL(T[j], j)), 7);
+		SS2 = SS1 ^ ROTL(A, 12);
+		TT1 = FF0(A, B, C) + D + SS2 + W1[j];
+		TT2 = GG0(E, F, G) + H + SS1 + W[j];
+		D = C;
+		C = ROTL(B, 9);
+		B = A;
+		A = TT1;
+		H = G;
+		G = ROTL(F, 19);
+		F = E;
+		E = P0(TT2);
+	}
+
+	for (j = 16; j < 64; j++) {
+		SS1 = ROTL((ROTL(A, 12) + E + ROTL(T[j], j)), 7);
+		SS2 = SS1 ^ ROTL(A, 12);
+		TT1 = FF1(A, B, C) + D + SS2 + W1[j];
+		TT2 = GG1(E, F, G) + H + SS1 + W[j];
+		D = C;
+		C = ROTL(B, 9);
+		B = A;
+		A = TT1;
+		H = G;
+		G = ROTL(F, 19);
+		F = E;
+		E = P0(TT2);
+	}
+
+	ctx->state[0] ^= A;
+	ctx->state[1] ^= B;
+	ctx->state[2] ^= C;
+	ctx->state[3] ^= D;
+	ctx->state[4] ^= E;
+	ctx->state[5] ^= F;
+	ctx->state[6] ^= G;
+	ctx->state[7] ^= H;
+}
+
+void sm3_update(struct sm3_context *ctx, const uint8_t *input, size_t ilen)
+{
+	size_t fill;
+	size_t left;
+
+	if (!ilen)
+		return;
+
+	left = ctx->total[0] & 0x3F;
+	fill = 64 - left;
+
+	ctx->total[0] += ilen;
+	ctx->total[0] &= 0xFFFFFFFF;
```

Comment:
`UINT32_MAX` ?
Since `total[]` is a `uint32_t` array, this mask seems useless.

Reply:
I'll remove this line